### PR TITLE
track schemes by name for error reporting

### PR DESCRIPTION
**Pull request by deads2k**
Getting an error message about a type not being in the scheme is hard to fix if you don't know which scheme is failing.  This adds a name to the scheme which can be set during creation or can be set based on the calling stack.  If you use the old constructor a name is generated for you based on the stack.  Something like "k8s.io/client-go/dynamic/scheme.go:28" for instance.

Also moves a typer to its point of use.  This was debt from previous refactors which I noticed going through.

@kubernetes/sig-api-machinery-misc 
@sttts 

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 1]
```diff
@@ -79,13 +79,22 @@ type Scheme struct {
 
 	// observedVersions keeps track of the order we've seen versions during type registration
 	observedVersions []schema.GroupVersion
+
+	// schemeName is the name of this scheme.  If you don't specify a name, the stack of the NewScheme caller will be used.
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: given how much copy/pasta of packages there is, I think always capturing the call site would be useful

**Comment by liggitt** [comment]
big +1 for having this info available. the number of times I've instrumented this locally to track something down is unreasonably large...

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 2]
```diff
@@ -127,73 +125,6 @@ func makeValidPrometheusMetricLabel(in string) string {
 // call chains to NewReflector, so they'd be low entropy names for reflectors
 var internalPackages = []string{"client-go/tools/cache/", "/runtime/asm_"}
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: can we have `"/runtime/asm_"` somewhere centrally and only append to that locally?

File: pkg/api/legacyscheme/scheme.go [hunk 3]
```diff
@@ -26,7 +26,7 @@ import (
 // extensions group instead. This Scheme is special and should appear ONLY in
 // the api group, unless you really know what you're doing.
 // TODO(lavalamp): make the above error impossible.
-var Scheme = runtime.NewScheme()
+var Scheme = runtime.NewNamedScheme("LegacyScheme")
```
Comment (sttts) [inline pkg/api/legacyscheme/scheme.go]: do we really want these strings? Or just use the filename from the strack everywhere? Wouldn't the later be enough? Would give less ugliness and less chances for error on copy&paste.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 2 above)
Comment (deads2k) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: > can we have "/runtime/asm_" somewhere centrally and only append to that locally?

done

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go (hunk 1 above)
Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: > given how much copy/pasta of packages there is, I think always capturing the call site would be useful

Done.  Such a lack of trust.

File: staging/src/k8s.io/apimachinery/pkg/util/naming/from_stack.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+		file = trimPackagePrefix(file)
+		name = fmt.Sprintf("%s:%d", file, line)
+		break
+	}
+	return name
+}
+
+// hasPackage returns true if the file is in one of the ignored packages.
+func hasPackage(file string, ignoredPackages []string) bool {
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/util/naming/from_stack.go]: A test would be nice

Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/util/naming/from_stack.go]: > A test would be nice

done

**Comment by deads2k** [comment]
boiler plate update.  retagging

**Pull request merged**